packages/core/src/errorCorrection/Polinomial/index.ts

```diff
@@ -0,0 +1,147 @@
+import { antilogAlpha, logAlpha } from '../alpha'
+
+type CoefficientExponentObject = {
+  [power: number]: number
+}
+
+const fromNumberArrayToExponentObject = (coefficients: number[]) => {
+  const exponentObject: CoefficientExponentObject = {}
+
+  coefficients.forEach((coefficient, index) => {
+    // We skip zero coefficients, as these can't be represented in alpha notation
+    if (coefficient === 0) return
+
+    exponentObject[index] = logAlpha(coefficient)
+  })
+
+  return exponentObject
+}
+
+const removeTrailingZeroes = (coefficients: number[]) => {
+  const lastNonZeroIndex =
+    coefficients.length -
+    coefficients
+      .slice()
+      .reverse()
+      .findIndex(coef => coef !== 0)
+
+  return coefficients.slice(0, lastNonZeroIndex)
+}
+
+const fromExponentObjectToNumberArray = (
+  exponents: CoefficientExponentObject
+) => {
+  const maxPower = Math.max(...Object.keys(exponents).map(Number))
+  const coefficients = Array(maxPower + 1).fill(0)
+
+  for (const [power, exponent] of Object.entries(exponents)) {
+    coefficients[Number(power)] = antilogAlpha(exponent)
+  }
+
+  return removeTrailingZeroes(coefficients)
+}
+
+export class Polinomial {
+  private coefficients: number[]
+  // We can use the alpha exponents of the coefficients for ease of multiplication
+  private exponents: CoefficientExponentObject
+
+  constructor(coefficients?: number[], exponents?: CoefficientExponentObject) {
+    if (coefficients) {
+      this.coefficients = removeTrailingZeroes(coefficients)
+      this.exponents = fromNumberArrayToExponentObject(coefficients)
+    } else if (exponents) {
+      this.coefficients = fromExponentObjectToNumberArray(exponents)
+      this.exponents = exponents
+    } else {
+      this.coefficients = []
+      this.exponents = {}
+    }
+  }
+
+  static fromCoef(coefficients: number[]) {
+    return new Polinomial(coefficients, undefined)
+  }
+
+  static fromCoefExponents(exponents: CoefficientExponentObject) {
+    return new Polinomial(undefined, exponents)
+  }
+
+  coefficientAtPower(position: number) {
+    if (position >= this.coefficients.length) {
+      return 0
+    }
+    return this.coefficients[position]
+  }
+
+  /**
+   * Returns the biggest power that has a non-zero coefficient
+   */
+  greaterPower() {
+    return Math.max(...Object.keys(this.coefficients).map(Number))
```

Comment:
Since we remove trailing zeroes at initialisation, we should use this.coefficients.length - 1

Reply:
Fixed in 21e9899